payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java

```diff
@@ -459,7 +460,37 @@ private PriorPaymentControlResult getPluginCreditResult(final PaymentControlCont
         if (prop == null) {
             return ImmutableMap.<UUID, BigDecimal>of();
         }
-        return (Map<UUID, BigDecimal>) prop.getValue();
+        // The deserialization may not recreate the map we expect, i.e Map<UUID, BigDecimal>, so we convert each key/value by hand
+        // See https://github.com/killbill/killbill/issues/1453
+        final Map<UUID, BigDecimal> res = new HashMap<>();
+        final Map m = (Map) prop.getValue();
+        for (final Object k : m.keySet()) {
+            UUID uuid;
+            if (k instanceof String) {
+                uuid = UUID.fromString((String) k);
+            } else if (k instanceof UUID) {
+                uuid = (UUID) k;
+            } else {
+                throw new IllegalStateException(String.format("Failed to deserialize plugin property map for adjustments: Invalid format for UUID, type=%s", k.getClass().getName()));
+            }
+
+            final Object v = m.get(k);
+            BigDecimal val;
+            if (v instanceof BigDecimal) {
+                val = (BigDecimal) v;
+            } else if (v instanceof String) {
+                val = new BigDecimal((String) v);
+            } else if (v instanceof Integer) {
+                val = BigDecimal.valueOf((Integer) v);
```

Comment:
Not sure this does what we want. Would it be easier to do something like?
```
try {
  val = new BigDecimal(v.toString());
} catch () { // throw }
```

Reply:
Actually, if the amount is an integer, it gets deserialized as an `Integer` and so I need this `else if` for this.

Reply:
Fixed in https://github.com/killbill/killbill/pull/1456/commits/a915db3c09ad7c6c384f7c3fe8e0848f36eda12b